essencium-backend/src/main/java/de/frachtwerk/essencium/backend/security/oauth2/OAuth2SuccessHandler.java

```diff
@@ -90,67 +90,75 @@ public void onAuthenticationSuccess(
       redirectHandler.setDefaultTargetUrl(oAuth2ConfigProperties.getDefaultRedirectUrl());
     }
 
-    if (!(authentication instanceof OAuth2AuthenticationToken)) {
-      LOGGER.error(
-          "did not receive an instance of {}, aborting",
-          OAuth2AuthenticationToken.class.getSimpleName());
-      redirectHandler.onAuthenticationSuccess(request, response, authentication);
-      return;
-    }
-
-    final var providerName =
-        ((OAuth2AuthenticationToken) authentication).getAuthorizedClientRegistrationId();
-
-    UserInfoEssentials userInfo;
-    try {
-      LOGGER.info(
-          "attempting to log in oauth2 user '{}' using provider '{}'",
-          authentication.getName(),
-          providerName);
-      userInfo = extractUserInfo((OAuth2AuthenticationToken) authentication, providerName);
-    } catch (UserEssentialsException e) {
-      LOGGER.error(e.getMessage());
-      redirectHandler.onAuthenticationSuccess(request, response, authentication);
-      return;
-    }
-
-    try {
-      final var user = userService.loadUserByUsername(userInfo.getUsername());
-      LOGGER.info("got successful oauth login for {}", userInfo.getUsername());
-
-      HashMap<String, Object> patch = new HashMap<>();
-
-      patch.put("firstName", userInfo.getFirstName());
-      patch.put("lastName", userInfo.getLastName());
-
-      if (oAuth2ConfigProperties.isUpdateRole()) {
-        List<Role> roles =
-            extractUserRole(((OAuth2AuthenticationToken) authentication).getPrincipal());
-        Role defaultRole = roleService.getDefaultRole();
-        if (roles.isEmpty() && Objects.nonNull(defaultRole)) {
-          LOGGER.info("no roles found for user '{}'. Using default Role.", userInfo.getUsername());
-          roles.add(defaultRole);
-        }
-        patch.put("roles", roles);
+    if (authentication instanceof OAuth2AuthenticationToken oAuth2AuthenticationToken) {
+      final String providerName = oAuth2AuthenticationToken.getAuthorizedClientRegistrationId();
+
+      final OAuth2ClientRegistrationProperties.ClientProvider clientProvider =
+          oAuth2ClientRegistrationProperties.getProvider().get(providerName);
+
+      UserInfoEssentials userInfo;
+      try {
+        LOGGER.info(
+            "attempting to log in oauth2 user '{}' using provider '{}'",
+            authentication.getName(),
+            providerName);
+        userInfo = extractUserInfo(oAuth2AuthenticationToken, providerName);
+      } catch (UserEssentialsException e) {
+        LOGGER.error(e.getMessage());
+        redirectHandler.onAuthenticationSuccess(request, response, authentication);
+        return;
       }
 
-      userService.patch(Objects.requireNonNull(user.getId()), patch);
-
-      redirectHandler.setToken(tokenService.createToken(user, SessionTokenType.ACCESS, null, null));
-    } catch (UsernameNotFoundException e) {
-      LOGGER.info("user {} not found locally", userInfo.getUsername());
+      try {
```

Comment:
The `try-and-catch` block should only be around the relevant code lines. For here only the `userService.loadUserByUsername(userInfo.getUsername());` can throw the `UsernameNotFoundException`. For a better readability, only this line should be covered by the try-and-catch block. The two possible flows (found vs. not found) can also be capsulated in two separate methods.